google/cloud/logging_v2/logger.py

```diff
@@ -133,6 +135,20 @@ def _do_log(self, client, _entry_class, payload=None, **kw):
         kw["labels"] = kw.pop("labels", self.labels)
         kw["resource"] = kw.pop("resource", self.default_resource)
 
+        severity = kw.get("severity", None)
+        if isinstance(severity, str) and not severity.isupper():
+            # convert severity to upper case, as expected by enum definition
+            kw["severity"] = severity.upper()
+
+        if isinstance(kw["resource"], collections.abc.Mapping):
+            # if resource was passed as a dict, attempt to parse it into a
+            # Resource object
+            try:
+                kw["resource"] = Resource(**kw["resource"])
+            except TypeError as e:
+                # dict couldn't be parsed as a Resource
+                raise TypeError(f"invalid resource dict. {e}")
```

Comment:
Rebinding the exception loses information. The default traceback should indicate where the failure occurred, and should be fine for the developer to debug:
```suggestion
            kw["resource"] = Resource(**kw["resource"])
```

Reply:
The problem is the raw error message doesn't make it clear that the resource is the issue:

`TypeError: __new__() missing 2 required positional arguments: 'type' and 'labels'`

That's a good point that we shouldn't lose the context with a fresh exception though. What do you think about [following this pattern](https://stackoverflow.com/a/46091127), to add the extra context?

```
try:
    kw["resource"] = Resource(**kw["resource"])
except TypeError as e:
    raise TypeError("invalid resource dict") from e
```